library/arith.lisp

```diff
@@ -480,6 +493,136 @@ The fields are defined as follows:
       (lisp Double-Float (x y)
         (cl:coerce (cl:/ x y) 'cl:double-float)))))
 
+(coalton-toplevel
+  (define-class (Trigonometric :a)
+    (sin   (:a -> :a))
+    (cos   (:a -> :a))
+    (tan   (:a -> :a))
+    (sinh  (:a -> :a))
```

Comment:
all of the hyperbolic functions are "just" different combinations of `(exp x)`. Can we just `define` them and have them constrained on `Exponentiable :t => :t -> :t`? We can `specialize` if we want to call the native Common Lisp variant.

To me, a smaller type class is better, if it's reasonable.

The big question, though, would be whether you want to define `tan` as a functon instead of just

```
(define (tan x)
  (/ (sin x) (cos x)))
```

I personally think it's OK for `tan` to be in the typeclass since you'll want `atan` there too...

Reply:
To your first point, I imagine it would make sense to implement the inverse hyperbolic functions outside the typeclass as well. I'll remove them and add them as functions.

One could argue that a reason that `tan` shouldn't be in the typeclass while `atan` should is because one can be computed "exactly" in terms of other methods in the typeclass (e.g. all implementations would have either `(/ (sin x) (cos x))` or an optimized version thereof for that method), while the other can't be (to my knowledge). In theory the only people who would be exposed to this idiosyncrasy would be people implementing the typeclass, for whom this would mean less work.

This is total bikeshedding, so feel free to tell me to go one way or the other and I'd oblige, but I think it makes sense to keep it outside the typeclass for now.

Reply:
I'm fine either way.